readthedocs/projects/tasks/utils.py

```diff
@@ -154,6 +160,81 @@ def send_external_build_status(version_type, build_pk, commit, status):
         send_build_status.delay(build_pk, commit, status)
 
 
+class DeprecatedConfigFileSiteNotification(SiteNotification):
+
+    failure_message = (
+        "Your project '{{ object.slug }}' doesn't have a "
+        '<a href="https://docs.readthedocs.io/en/stable/config-file/v2.html">.readthedocs.yaml</a> '
+        "configuration file. "
+        "This feature is <strong>deprecated and will be removed soon</strong>. "
+        "Make sure to create one for your project to keep your builds working."
+    )
+    failure_level = WARNING_PERSISTENT
+
+
+class DeprecatedConfigFileEmailNotification(Notification):
+
+    app_templates = "projects"
+    name = "deprecated_config_file_used"
+    context_object_name = "project"
+    subject = "Your project will start failing soon"
+    level = REQUIREMENT
+
+    def send(self):
+        """Method overwritten to remove on-site backend."""
+        backend = EmailBackend(self.request)
+        backend.send(self)
+
+
+@app.task(queue="web")
+def deprecated_config_file_used_notification(build_pk):
```

Comment:
I assumed that we'd be implementing this as a scheduled task that looped over all projects to check for a build with a config file or something. Otherwise, we're only going to reach projects that have recently built their project.

Obviously build activity is a good measure of how much we care about notifying the project, but I'd also say that projects aren't actively building should be on our radar as well.

It would also remove the need for caching in this task, as we could just use a scheduled celery task. I'm guessing there are different hurdles there, like what we've hit in the past with iterating over a large number of model rows in scheduled tasks.

Reply:
Hrm, I'm not sure how to do this in a effective and performant way. Definitely, we can't iterate over all the projects and check if they have build without a config file in one task without caching. We are doing something similar for spam, where we iterate over N projects per task and we save the `id` of the last project we check. The next time the task is executed we continue from there.

> we're only going to reach projects that have recently built their project.

I think this is a good start for now while we think about how to reach the others.

Reply:
I think I found a query that could work here to do a scheduled Celery task once a week:

```python
In [23]: projects = set()
    ...: # NOTE: we could skip the projects with a spam score > 150,
    ...: # to reduce the amount of email/onsite notifications we send
    ...: for p in Project.objects.all().only('id'):
    ...:     # NOTE: intead of iterating over all the active versions,
    ...:     # we can only consider the default one
    ...:     for v in p.versions.filter(active=True).only('id'):
    ...:         b = v.builds.filter(success=True).only('_config').first()
    ...:         if b and b.deprecated_config_used():
    ...:             projects.add(p.slug)
    ...:             # Do not continue iterating over the
    ...:             # other versions when we know this project
    ...:             # will get the notification
    ...:             break

In [24]: projects
Out[24]: {'sphinxcontrib-bibtex', 'test-builds', 'tutorial-template'}
```

I'm using `.only` to retrieve as few as possible fields from the row. Filtering by `Version(active=)` and `Build(success=)` I think it will help a lot here, but I haven't tested the query on production yet. I need to solve my SSH connection that stopped working for some reason now 🙃

Reply:
I updated the code to use a scheduled Celery task. I say we should run it manually first before deploying this code to production to check what's gonna happen once deployed.